File: smart_travel_FE/lib/domain/entities/tour.dart
class Tour{
  final int id; // BẮT BUỘC, không null
  final String? name;
  final int? destinationId; // Chỉ lưu ID thay vì object
  final String? description;
  final int? durationDays;
  final int? durationNights;
  final double? pricePerPerson;
  final int? maxPeople;
  final int? minPeople;
  final Map<String, dynamic>? included; // JSON → Map
  final Map<String, dynamic>? excluded; // JSON → Map
  final double? averageRating;
  final int? reviewCount;
  final int? bookingCount;
  final bool? isActive;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final List<String>? imageUrls; // Danh sách URL ảnh
  final String? image;

  Tour({
    required this.id,
    this.name,
    this.destinationId,
    this.description,
    this.durationDays,
    this.durationNights,
    this.pricePerPerson,
    this.maxPeople,
    this.minPeople,
    this.included,
    this.excluded,
    this.averageRating,
    this.reviewCount,
    this.bookingCount,
    this.isActive,
    this.createdAt,
    this.updatedAt,
    this.imageUrls,
    this.image,
  });
}

File: smart_travel_BE/src/main/java/com/example/smart_travel_BE/entity/Tour.java
package com.example.smart_travel_BE.entity;

import jakarta.persistence.*;
import lombok.Data;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Data
@Entity
@Table(name = "tours",
        indexes = {
                @Index(name = "idx_destination_id", columnList = "destination_id"),
                @Index(name = "idx_price", columnList = "price_per_person")
        })
public class Tour {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "name", nullable = false, length = 200)
    private String name;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "destination_id", nullable = false)
    private Destination destination;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "duration_days", nullable = false)
    private Integer durationDays;

    @Column(name = "duration_nights", nullable = false)
    private Integer durationNights;

    @Column(name = "price_per_person", nullable = false, precision = 12, scale = 2)
    private BigDecimal pricePerPerson;

    @Column(name = "max_people", nullable = false)
    private Integer maxPeople;

    @Column(name = "min_people", nullable = false)
    private Integer minPeople = 1;

    @Column(name = "included", columnDefinition = "JSON")
    private String included;

    @Column(name = "excluded", columnDefinition = "JSON")
    private String excluded;

    @Column(name = "average_rating", precision = 3, scale = 2)
    private BigDecimal averageRating = BigDecimal.ZERO;

    @Column(name = "review_count", nullable = false)
    private Integer reviewCount = 0;

    @Column(name = "booking_count", nullable = false)
    private Integer bookingCount = 0;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private User owner;

    @OneToMany(mappedBy = "tour", cascade = CascadeType.ALL, orphanRemoval = true)
    @Fetch(FetchMode.SUBSELECT)
    private List<TourSchedule> schedules = new ArrayList<>();

    @OneToMany(mappedBy = "tour", cascade = CascadeType.ALL, orphanRemoval = true)
    @Fetch(FetchMode.SUBSELECT)
    private List<TourImage> images = new ArrayList<>();

}